### ISSUE TITLE
ServletOutputStream.isReady() returns true while it is not yet ready

### ISSUE THREAD

**Issue by cy6erGn0m** [eclipse/jetty.project#1755]
The general contract for first write is the following
- set write listener via `setWriteListener(...)`
- suspend and await for `onWritePossible()` invocation
- check for `isReady()` and if it returned `true` then `write(..)` 

However `isReady()` just after write listener has been set is `true`. Shouldn't it be `false` until first `onWritePossible()` callback invocation? Or it should be possible to invoke `write` if `isRead()` returned `true` in spite of that write listener will be scheduled later. I'd say the first option looks better

**Comment by sbordet** [comment]
Method `isReady()` is not a method that should be called lightly. It's not a query method that returns a field.

It's a method that, when it returns false, has the side effect of telling the container to schedule a call to `onWritePossible()`.

As such, calling `isReady()` from `service()` (or `doGet()`, etc.) is not a good idea.

Is it correct that it returns `true`, however, because you _can_ write to the `OutputStream` if you want to, and the container will invoke `onWritePossible()` the first time without the need for `isReady()` to be invoked by the application (and returning `false`).

Yes, the API is weird.

**Comment by gregw** [comment]
initially writing is always possible as the only thing that can prevent writing is if a prior write has not completed. so onWritePossible is scheduled as soon as a listener is set.

even if not possible, the code should always check isReady before writing so a spurious call to onWritePossible should not be a problem.

**Comment by cy6erGn0m** [comment]
Just note that the following will crash in `write(...)`

```
out.setWriteListener(..)
if (out.isReady()) { out.write( ... ) }
```

**Comment by sbordet** [comment]
"Crash" ? Do you have details ?

**Comment by cy6erGn0m** [comment]
`IllegalStateException`. Let me try to find where exactly

**Comment by cy6erGn0m** [comment]
```
java.lang.IllegalStateException: null
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:455)
	.....
```

at attempt to switch state

```java
if (!_state.compareAndSet(OutputState.PENDING, OutputState.ASYNC))
    throw new IllegalStateException();
```

**Comment by sbordet** [comment]
I'm not sure that is legal usage of the Servlet API, but I'll defer to @gregw on that.

But what's the point of that code ?
If you set a `WriteListener`, why you want to try to write the next line ?
Just put the next line inside `onWritePossible()` and then container will call you.

**Comment by cy6erGn0m** [comment]
Yeah, I know that the correct way is to wait for `onWritePossible()` but I was forced to do so to workaround tomcat's bug that sometimes doesn't invoke callback at all for upgraded output stream. So I finally ended up with a temporary additional check for ServletOutputStream class name :(

Actually I am facing issues with async Servlet API because the specification is too often not clear about threading and usage scenarios

**Comment by sbordet** [comment]
Well, I guess we can avoid that exception, so let me reopen this issue.

**Comment by gregw** [comment]
The problem is that code like `if (out.isReady()) { out.write( ... ) }` is a race condition if you don't respect the servlet containers threading.   If any thread can execute that code while a call to `onWritePossible` might occur, then you will have a race between the test and the write and one thread may fail.

So I'm not really inclined to try to "fix" this because I'm not sure how to fix it - specially if the `setWriteListener` is called from a non container thread and thus the `onWritePossible` call will be dispatched and in a race.

**Comment by sbordet** [comment]
@gregw, however this may be legal:

```java
protected void service(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
    AsyncContext context = request.startAsync();
    ServletOutputStream out = response.getOutputStream();
    out.setWriteListener(...);
    while (out.isReady()) {
        out.write(...);
    }
}
```
Above, we don't exit `service()` until we have written it all, or we're not ready.
In case of the latter, `onWritePossible()` will be called.

I recognize that there is code duplication, as you have to write the same code _after_ `setWriteListener()`, and inside `onWritePossible()`, and so would be better to write it once in `onWritePossible()`. But still, seems legal code to me ?

**Comment by cy6erGn0m** [comment]
Of course interference between "main" thread and one that will be running `onWritePossible` callback is potentially dangerous so it is the user's responsibility to avoid concurrent write loops. It is complicated, dangerous and error-prone but I believe it should be some clear position about this particular scenario: it should be either allowed or strictly prohibited.

**Comment by cy6erGn0m** [comment]
I personally do such a synchronization so write loop is spinning while possible on caller thread (could be any thread, possibly container's processor thread) and then moving to `onWritePossible()`'s thread if `isReady()` return false but of course write loops are never running concurrently on different threads.

**Comment by gregw** [comment]
@sbordet

I'm not convinced.

Once setWriteListener is called then a callback to owp is scheduled as per
the spec.

If we then allow an isReady write loop what do we do if false is eventually
returned?  Cancel the owp until a write is possible, or call it once or
call it twice.  All contravene the spec in some way.

**Comment by joakime** [comment]
@gregw and @sbordet this issue needs an update (or should be closed)
